---
title: "Final Assignment"
author: "Alejandro Roca"
date: "2017 M11 30"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(tidyverse)
```

# Whole-genome patterns of nucleotide diversity & recombination in the model legume Medicago truncatula

## 1. Import and filter the dataset in R.  

*Read about the metadata: how is the data organized, what do variables in the data mean roughly.  How many chromosomes, how many genomic windows are available in the dataset? Are obvious windows that are “big” outliers?*

*Choose a subset of the dataset by excluding windows that have very few SNPs (LdH_SNPs) and/or a window size that is very small (examine the distribution of ldH_win_size).*

```{r}
data <- read.delim("Data_Mt_DSBionf.csv", sep = ",")
data$chr <- as.factor(data$chr)
plot_data <- data %>% ggplot()
plot_data + geom_histogram(mapping = aes(x = LdH_SNPs, ..density..), bins = 50) + 
    labs(title = "Number of SNPs per window distribution", x = "SNP count", y = "Density")
plot_data + geom_histogram(mapping = aes(x = ldH_win_size, ..density..), bins = 100) +
    labs(title = "Window size distribution", x = "Window size (bp)", 
         y = "Density")
```

Looking at the distribution we can filter the data to retain the ones within an interval of 3 times the sd from the mean.

```{r}
#Filtering for number of SNPs
lowlim <- mean(data$LdH_SNPs) - 3*sd(data$LdH_SNPs)
highlim <- mean(data$LdH_SNPs) + 3*sd(data$LdH_SNPs)
d <- data %>%
    filter(LdH_SNPs > lowlim) %>%
    filter(LdH_SNPs < highlim)

#Filtering for window size
lowlim <- mean(data$ldH_win_size) - 3*sd(data$ldH_win_size)
highlim <- mean(data$ldH_win_size) + 3*sd(data$ldH_win_size)
d <- d %>%
    filter(ldH_win_size > lowlim) %>%
    filter(ldH_win_size < highlim)
```

*Check what percentage of the windows you keep / discard.*

```{r}
cat("We keep ", round(dim(d)[1] / dim(data)[1] * 100, digits = 2), "% of the data\n", sep = "")
```

*Keep that filtered subset fixed for the remaining analysis.*

---------------------------------------------------------------------------

## 2. Displaying genomic data

*Make a graph that illustrates patterns of variation in recombination (as measured by rho) and polymorphism (as measured by qp.site) along chromosome 5 (see for instance Fig 4 of the paper).*

```{r}
plot_chr5 <- d %>% filter(chr == "5") %>% ggplot()

plot_chr5 +
    geom_area(mapping = aes(y = rho_kb / rho_theta / 1000, x = dist_cent), fill = "royalblue4") +
    geom_area(mapping = aes(y = rho_kb / 1000, x = dist_cent), fill = "red3") +
    geom_vline(xintercept = 0) +
    labs(title = "Chromosome 5", x = "Distance to centromere (bp)", 
         y = expression(paste(theta, "w (blue) and ", rho, " (red)")))

plot_chr5 + 
    geom_line(mapping = aes(y = rho_theta, x = dist_cent)) +
    geom_vline(xintercept = 0, color = "red") +
    geom_hline(yintercept = 1, linetype = "dashed") +
    labs(title = "Chromosome 5", x = "Distance to centromere (bp)", 
     y = expression(paste(rho, "/", theta, "w")))

plot_chr5 +
    geom_line(mapping = aes(y = gene_dens, x = dist_cent), color = "forestgreen") +
    geom_vline(xintercept = 0, color = "red") +
    labs(title = "Chromosome 5", x = "Distance to centromere (bp)", 
     y = "Gene Density")
```


## 3. Summarizing levels of Recombination and polymorphism in M. truncatula

*Use a boxplot or any other graphically more convenient way to illustrate the distribution of nucleotide diversity (qp.site) and recombination rate (rho per kb) among the different chromosomes of M. truncatula.*

```{r}
d %>% ggplot(mapping = aes(x = chr, y = qp.site)) +
    geom_violin() +
    geom_boxplot(width=0.1) +
    stat_summary(fun.y=mean, geom="point", size=2, color = "red4") +
    labs(title = "Nucleotide diversity distribution by chromosome", 
         x = "Chromosome", y = "Mean nucleotide diversity per site")

d %>% ggplot(mapping = aes(x = chr, y = rho_kb)) +
    geom_violin() +
    geom_boxplot(width=0.1) +
    stat_summary(fun.y=mean, geom="point", size=2, color = "red4") +
    labs(title = "Recombination rate distribution by chromosome", 
         x = "Chromosome", y = "Recombination rate (ρ) per kb")
```

*Are means or median better summaries for these variables?*

As we can see in the plots, although there are higher bound outliers which drag the mean to higher values, this phenomenom is not very strong. Therefore, we can take the mean as a good summary for these variables.

*For each chromosome calculate the (mean) (or median) recombination rate and its associated 95% CI. Present these results as a table.*  

We will calculate the Confidence Interval by bootstrapping. We will resample the data n times and calculate the mean for each resampling. Then, with the distribution obtained from the bootstrap of means we will check the 0.025 quantile and the 0.975 quantile to obtain the lower and upper boundary of the CI respectively.

```{r}
d_chr <- d %>% group_by(chr) %>% 
    summarise(median_rho = median(rho_kb),
              mean_rho = mean(rho_kb)) #Calculate recombination rate mean by chromosome

N = 1000
boot_mean <- as.data.frame(matrix(data = 0, nrow = N, ncol = length(levels(d$chr))))
colnames(boot_mean) <- levels(d$chr)
for (i in levels(d$chr)) {
    d_chr_i <- d %>% filter(chr == i)
    for (n in 1:N) {
        boot_mean[n, i] <- mean(sample_n(d_chr_i, dim(d_chr_i)[1], replace = T)$rho_kb)
    }
}

CI <- t(apply(boot_mean, 2, function(x) quantile(x, c(0.025, 0.975))))
means <- apply(boot_mean, 2, mean)

cbind(d_chr[,c(3)], CI)
```

*The authors state that:*

>“The population-scaled recombination rate is approximately one-third of the mutation rate, consistent with expectations for a species with a high selfing rate.” 

*Using the ratio calculated by the authors (rho_theta) discuss if you support that assertion.*  

Being this statement true, the distribution ratio $\rho/\theta$ should have a mean of $1/3$. Visually:

```{r}
d %>% ggplot() +
    geom_histogram(aes(x = rho_theta, ..density..), bins = 500) +
    geom_vline(xintercept = 1/3, color = "red") +
    geom_text(label = "1/3", x = 0.5, y = 2.5, color = "red")
```


As we have a big sample, a good way of checking if this statement is true would be to make a Confidence Interval for the mean by bootstrap. We will then check whether the value 1/3 falls inside the CI.

```{r}
N = 1000
boot_mean <- c()
for (i in levels(d$chr)) {
    boot_mean[i] <- mean(sample(d$rho_theta, length(d$rho_theta), replace = T))
}

quantile(boot_mean, c(0.025, 0.975))
```

According to this, we can't say that the distribution of the ratio $\rho/\theta$ has a mean equal to $1/3$.

Other tests regarding goodness of fit between the distribution of $\rho$ and $theta/3$, or one sample t-test for the log(rho_theta) (being H0: $\mu = log(1/3)$) lead to the same conclusion.

*Is there a lot of variation between chromosomes for this ratio?*  

We can answer this question performing an ANOVA test.

```{r}
lm <- lm(rho_theta ~ chr, data = d)
anova(lm)
```

As the Analysis of Variance Table returns, there is a lot of variation between chromosomes for this ratio. It may happen that some chromosomes follow this hypothesis and some don't so we will perform the above analysis for each chromosome.

```{r}
d_chr <- d %>% group_by(chr) %>% 
    summarise(mean_rho_theta = mean(rho_theta))

N = 1000
boot_mean <- as.data.frame(matrix(data = 0, nrow = N, ncol = length(levels(d$chr))))
colnames(boot_mean) <- levels(d$chr)
for (i in levels(d$chr)) {
    d_chr_i <- d %>% filter(chr == i)
    for (n in 1:N) {
        boot_mean[n, i] <- mean(sample_n(d_chr_i, dim(d_chr_i)[1], replace = T)$rho_theta)
    }
}

CI <- t(apply(boot_mean, 2, function(x) quantile(x, c(0.025, 0.975))))
cont <- ifelse(CI[, 1] <= 1/3 & CI[, 2] >= 0.3, "*", ".") # '*': Yes, '.': No

data.frame(CI_left = CI[,1], CI_right = CI[, 2], match_hypothesis = cont)
```

We can see that the statement made by the authors is true for chromosomes 3 and 5.

## 4. A test of an evolutionary hypothesis

*One of the conclusion of the paper made by the author is that:*  

>“Nucleotide diversity in 100-kb windows was negatively correlated with gene density, which is expected if diversity is shaped by selection acting against slightly deleterious mutations.“ 

*We want to reproduce analysis and discuss if our analysis support this conclusion?*   

```{r}
cor.test(x = d$gene_dens, y = d$qp.site)

d %>% ggplot(aes(x = gene_dens, y = qp.site)) +
    geom_point() +
    geom_smooth(method = "lm")

d %>% ggplot(aes(x = gene_dens, y = qp.site)) +
    facet_wrap(~chr) +
    geom_point() +
    geom_smooth(method = "lm")
```

From the Pearson's correlation test we can see that there is indeed a significant negative correlation between the nucleotide diversity in 100-kb windows (`qp.site` column in data) and the gene density (`gene_dens`). This is also shown in both genome-wide and per chromosome plots.

*The so called “background selection hypothesis” states that genomic regions that are experiencing more deleterious mutations may exhibit overall less polymorphism because selection as it “weeds out” delterious mutations also eliminates other SNP variation in their vicinity. This hypothesis makes the prediction that “everything else being equal”:*  

- *Gene-dense regions (because they are more prone to produce deleterious mutation) should be exhibiting comparatively less polymorphism than regions that are gene poor.*  
- *Background selection will have more influence in regions that have a low recombination rate.*  

*So far testing this hypothesis and its predictions was not easy (except for a few model species) because it requires enormous amounts of polymorphism data. We want to test that hypothesis in M truncatula.*  

**4.1 Formulate expectations for what genomic covariate will explain variation in nucleotide diversity (qp.site) of a window.**

First of all, based on the "background selection hypothesis" and the correlation between nucleotide diversity and gene density, we would expect that the __gene density__ will explain variation in nucleotide diversity. Also, the __recombination rate__ is expected to play an important role in the nucleotide diversity. Moreover, we can see there is a high variation between chromosomes for the `qp.site` variable so we could hypothesize that the nucleotide diversity is also dependant on the __chromosome__. Finally, we could also think that the nucleotide diversity depends on the __position__ of the chromosome.

Therefore, we will perform several correlation test for this variables to determine the best explanation for the variation in nucleotide diversity.

*Using linear models, explore if there is support for the predictions of the background selection hypothesis in the patterns of M truncatula genetic diversity.*  

By the background selection hypothesis, gene density and recombination rate should explain the nucleotide diversity.

```{r}
lm1 <- lm(data = d, qp.site ~ gene_dens)
lm2 <- lm(data = d, qp.site ~ gene_dens*rho_kb) # Hypothesis model
# summary(lm2) # interaction gene_dens:rho_kb seems to be important
anova(lm1, lm2) # gene_dens:rho_kb interaction improves the explanation of qp.site
```

**4.2 Are all chromosome of the M truncatula genome exhibiting similar patterns?**

*(Explore this visually and using appropriate linear models that include a chromosome factor and how chromosome can “interact” with the factors of your analysis)*  

```{r}
d %>% ggplot(aes(x = gene_dens, y = qp.site, color = chr)) +
    geom_smooth(method = "lm", se = F)

d %>% ggplot(aes(x = rho_kb, y = qp.site, color = chr)) +
    geom_smooth(method = "lm", se = F)
```

As we can see visually in the plots, the adjusted linear model for `qp.site ~ gene_dens` in each chromosome seem to have the same slope for every chromosome except for chromosome 4. Same exception happens when we represent `qp.site ~ rho_kb`, but this time with chromosomes 3 and 5. This could mean that in those chromosome there is an interaction between the chromosome number and the recombination rate/gene density.

If we add the `chr` variable to our model we get a better explanation for `qp.site`:

```{r}
lm3 <- lm(data = d, qp.site ~ gene_dens*rho_kb + chr)
anova(lm2, lm3) #adding the chr variable improve the explanation of qp.site
# summary(lm3)
```

*Can other covariate potentially confounding the background selection hypothesis predictions?*

As I mentioned above, we could think that for example the position in the chromosome influences the nucleotide diversity.

```{r}
lm4 <- lm(data = d, qp.site ~ gene_dens*rho_kb + chr + dist_cent)
anova(lm3, lm4) #adding the chr variable improve the explanation of qp.site
# summary(lm3)
```

We can see, `dist_cent` variable seems to be significant, although this could be because the recombination rate varies with the position in the genome as we saw before.

## 5. Tracking footprints of recent and intense selection in the Medicago truncatula genome.

**5.1 According to the so-called “selective sweep” scenario, if a region contains a mutation that was recently selected, this recent and intense selection is likely to wipe out most of the polymorphism in that region.**

*Do you find footprints for a recent major selective sweep as evidenced by 2 or more consecutive windows with very low diversity (qp.site) relative to the rest of the genome? Hint: explore graphically where the least polymorphic windows (use the 5% least polymorphic windows within the genome) are located.* 

```{r}
low.div <- quantile(d$qp.site, 0.05)
d_low.div <- d %>%
    filter(qp.site <= low.div)

d_low.div %>% ggplot(aes(y = qp.site, x = prop_dist, color = chr)) +
    geom_point()
```

Exploring the relative position of the least (5%) polymorphic windows, apparently there is a greater number of them located in the positions far from the centromere than in the other parts of the chromosomes. 

Answering the question:

```{r}
diff_X <- diff(d_low.div$X) #calculate differerence between window number (1 -> consecutive)
result <- rle(diff_X) #calculate number of times a value is repeated
max_cons <- max(result$lengths[result$values == 1]) + 1 # X differences between X + 1 elements (windows)

cat("The maximum number of consecutive windows with very low diversity is:", 
    max_cons)   
```

Yes, we can find footprints for a recent major selective sweep. There is one place where we can find 2 or more (7) consecutive windows. Concretely in chromosome 3:

```{r}
consec <- seq(which(result$lengths == max_cons - 1), by = 1, length.out = max_cons)
select(d_low.div[consec,], window, chr, qp.site)
```

**5.2 Is the location of least polymorphic windows randomly distributed on chromosomes?**

*Hint: To make such a test, you can group windows into larger segments comprising 20 windows and count the number of windows containing least polymorphic windows in each segment. “On average” you expect 1 of such 5% least polymorphic windows per segment. Think about what probability distribution you expect can capture the fact that windows with low level of polymorphisms are occurring “at random” along the chromosome in each segment. Then decide on a statistic you can use to test the null hypothesis that the “location of least polymorphic windows is randomly distributed along the chromosomes”. Get a null distribution for such test statistic and state whether you reject the null or not for each chromosome by discussing the p-values you obtain.*

We can address this problem in the following way:  
Considering the null hypothesis as a binomial distribution $X ~ B(n, p)$ with $p = 0.05$, as we know the number of "trials" (n) for each chromosome. Performing a binomial test will let us decide if the number of least polymorphic windows is different than 1 per group of 20 windows, which is the expected by chance under this conditions.   

```{r}
d_by_chr <- split(d, d$chr) #Divide data set by chromosomes

least.counter <- function(x){
    
    ### Divide in chunks ###
    chunk <- 20
    n <- nrow(x)
    r  <- rep(1:ceiling(n/chunk), each=chunk)[1:n]
    x <- split(x, r) #Split chromosome windows in groups of 20
    # x <- x[lapply(x, nrow) == 20] #Filter out (last) window with less elements
    
    ### Count number of least polymorphic windows per chunk ###
    sapply(x, function(y) nrow(filter(y, qp.site <= low.div)))
}

lcounts_by_chr <- lapply(d_by_chr, least.counter)

bin_tests <- lapply(lcounts_by_chr, function(x) binom.test(sum(x == 1), length(x)*20, p = 0.05))
data.frame(chr = names(bin_tests), random = sapply(bin_tests, function(x) x$p.value > 0.05))
```

With a 5% thresehold, the distribution of the least polymorphic sites only appears to be randomly distributed for the chromosome 4.

**5.3** *Another typical footprint of a selective sweep is, besides having lower polymorphism, the fact that genomic regions that have experienced a recent selective sweep are expected to have an excess of rare variants (i.e. the only mutations that contribute to polymorphism are recent and therefore still in low frequency). The excess of rare variants in a window can be measured through the summary statistics “Tajima’s D”. Negative Tajima’s D values mean an excess of rare variants (SNPs) while positive Tajima’s D means an excess of frequent SNPs relative to a neutrally evolving region.*

*Are windows exhibiting abnormally low amounts of polymorphism –as measured by qp.site are also having more negative Tajima’s D values?*

```{r warning=FALSE}
d %>% ggplot(aes(x = qp.site, y = TajimasD)) +
    geom_point(color = "grey") +
    geom_point(data = d_low.div, aes(x = qp.site, y = TajimasD)) +
    facet_wrap(~chr) +
    geom_smooth(data = d_low.div, aes(x = qp.site, y = TajimasD), method = "lm", 
                se = F, fullrange = T) +
    ylim(c(-3, 1))
```

Yes, we can see this is met for every chromosome except for the 4. This together with the previous results would make as think that chromosome 4 does not present a selective sweep footprint. 

*Use some statistical inference method to examine the relationship. Could other confounding variable obscure this relationship?*

```{r}
lm5 <- lm(data = d, qp.site ~ TajimasD)
# summary(lm5)
```

We can see there is a significant linear relationship between nucleotide diversity and the amount of rare variants. As we saw in the graphs, the least polymorphic the sites are, the more negative is Tajima's D, indicating an excess of rare SNPs.

Adding this variable to our null model defined above:

```{r}
lm6 <- lm(data = d, qp.site ~ gene_dens * rho_kb  + chr + TajimasD )
anova(lm3, lm6)
```

We can see there is an increased fit when we add Tajima's D to our model. 