---
title: "Final Assignment"
author: "Alejandro Roca"
date: "2017 M11 30"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(tidyverse)
```

# Whole-genome patterns of nucleotide diversity & recombination in the model legume Medicago truncatula

<!-- Notes for project:  -->

<!-- filter (maybe filter windows with small size) -->

<!-- finding CI for median: bootstrap approach (slice data by chromosome and resample them, calculate the median and get the distribution) - example in book (chimpanzee example) -->

<!-- 4- how much gen diversity vary among windows: qp.site ~ recomb (nt div per site, are the SNP far away or not?) whar are the variables to explain window diversity (qp.site) -> chr, recomb, genedensity   -->
<!-- *background selection: deleterious mutations are eliminated from population and that takes other SNPs with it (LD), we would expect less density in the window. Look how are compared models taking into account the gene density and models taking into account also chr, recombination.... Graph visualize faceting (by one variable/two variables at a time... choose). Models: -->
<!-- - qp.site ~ recomb -->
<!-- - qp.site ~ recomb + genedensity (are they interacting? enough to explain the data?) -->
<!-- - qp.site ~ recomb + chr + chr*genedensity -->
<!-- - qp.site ~ chr + recomb + genedensity (!declare chr as factor) -->
<!-- - qp.site ~ chr*recomb + genedensity -->

<!-- Check how the fit is incremented by comparing the models by anova() or type II sum of squares -->

<!-- 5.2 Do we hit the lower 5% randomly(?)? binomial: you know the events that we have and those events should be K p=0.05 (of being polimorfic) ~ poisson with lambda = K p. We can split our 1808 windows in 180 windows with k = 10 and this can be seen as many draws from a binomial distribution or a poisson distributed along windows. If k p is not too large this two distribution are very similar. -->

<!-- 5.3 do graph and see the level of (negative) correlation (it assumes that there is not other variables influencing). Then we can see if the Tajima's D is actually having an effect when we correct for our null model (qp.site ~ gene_dens + chr). In other words, is the Tajima's D explaining the qp.site or is the Tajima's D related to other variables that influence qp.site? -->

## 1. Import and filter the dataset in R.  

*Read about the metadata: how is the data organized, what do variables in the data mean roughly.  How many chromosomes, how many genomic windows are available in the dataset? Are obvious windows that are “big” outliers?*

*Choose a subset of the dataset by excluding windows that have very few SNPs (LdH_SNPs) and/or a window size that is very small (examine the distribution of ldH_win_size).*

```{r}
data <- read.delim("Data_Mt_DSBionf.csv", sep = ",")
data$chr <- as.factor(data$chr)
plot_data <- data %>% ggplot()
plot_data + geom_histogram(mapping = aes(x = LdH_SNPs, ..density..), bins = 50) + 
    labs(title = "Number of SNPs per window distribution", x = "SNP count", y = "Density")
plot_data + geom_histogram(mapping = aes(x = ldH_win_size, ..density..), bins = 100) +
    labs(title = "Window size distribution", x = "Window size (bp)", y = "Density")
```

Looking at the distribution we can filter the data to retain the ones within an interval of 3 times the sd from the mean.

```{r}
#Filtering for number of SNPs
lowlim <- mean(data$LdH_SNPs) - 3*sd(data$LdH_SNPs)
highlim <- mean(data$LdH_SNPs) + 3*sd(data$LdH_SNPs)
d <- data %>%
    filter(LdH_SNPs > lowlim) %>%
    filter(LdH_SNPs < highlim)

#Filtering for window size
lowlim <- mean(data$ldH_win_size) - 3*sd(data$ldH_win_size)
highlim <- mean(data$ldH_win_size) + 3*sd(data$ldH_win_size)
d <- d %>%
    filter(ldH_win_size > lowlim) %>%
    filter(ldH_win_size < highlim)
```

*Check what percentage of the windows you keep / discard.*

```{r}
cat("We keep ", round(dim(d)[1] / dim(data)[1] * 100, digits = 2), "% of the data\n", sep = "")
```

*Keep that filtered subset fixed for the remaining analysis.*

---------------------------------------------------------------------------

## 2. Displaying genomic data

*Make a graph that illustrates patterns of variation in recombination (as measured by rho) and polymorphism (as measured by qp.site) along chromosome 5 (see for instance Fig 4 of the paper).*

```{r}
plot_chr5 <- d %>% filter(chr == "5") %>% ggplot()

plot_chr5 +
    geom_area(mapping = aes(y = rho_kb / rho_theta / 1000, x = dist_cent), fill = "royalblue4") +
    geom_area(mapping = aes(y = rho_kb / 1000, x = dist_cent), fill = "red3") +
    geom_vline(xintercept = 0) +
    labs(title = "Chromosome 5", x = "Distance to centromere (bp)", 
         y = expression(paste(theta, "w (blue) and ", rho, " (red)")))

plot_chr5 + 
    geom_line(mapping = aes(y = rho_theta, x = dist_cent)) +
    geom_vline(xintercept = 0, color = "red") +
    geom_hline(yintercept = 1, linetype = "dashed") +
    labs(title = "Chromosome 5", x = "Distance to centromere (bp)", 
     y = expression(paste(rho, "/", theta, "w")))

plot_chr5 +
    geom_line(mapping = aes(y = gene_dens, x = dist_cent), color = "forestgreen") +
    geom_vline(xintercept = 0, color = "red") +
    labs(title = "Chromosome 5", x = "Distance to centromere (bp)", 
     y = "Gene Density")
```


## 3. Summarizing levels of Recombination and polymorphism in M. truncatula

*Use a boxplot or any other graphically more convenient way to illustrate the distribution of nucleotide diversity (qp.site) and recombination rate (rho per kb) among the different chromosomes of M. truncatula.*

```{r}
d %>% ggplot(mapping = aes(x = chr, y = qp.site)) +
    geom_violin() +
    geom_boxplot(width=0.1) +
    stat_summary(fun.y=mean, geom="point", size=2, color = "red4") +
    labs(title = "Nucleotide diversity distribution by chromosome", 
         x = "Chromosome", y = "Mean nucleotide diversity per site")

d %>% ggplot(mapping = aes(x = chr, y = rho_kb)) +
    geom_violin() +
    geom_boxplot(width=0.1) +
    stat_summary(fun.y=mean, geom="point", size=2, color = "red4") +
    labs(title = "Recombination rate distribution by chromosome", 
         x = "Chromosome", y = "Recombination rate (ρ) per kb")
```

*Are means or median better summaries for these variables?*

As we can see in the plots, although there are higher bound outliers which drag the mean to higher values, this phenomenom is not very strong. Therefore, we can take the mean as a good summary for these variables.

*For each chromosome calculate the (mean) (or median) recombination rate and its associated 95% CI. Present these results as a table.*  

We will calculate the Confidence Interval by bootstrapping. We will resample the data n times and calculate the mean for each resampling. Then, with the distribution obtained from the bootstrap of means we will check the 0.025 quantile and the 0.975 quantile to obtain the lower and upper boundary of the CI respectively.

```{r}
d_chr <- d %>% group_by(chr) %>% 
    summarise(median_rho = median(rho_kb),
              mean_rho = mean(rho_kb)) #Calculate recombination rate mean by chromosome

N = 100
boot_mean <- as.data.frame(matrix(data = 0, nrow = N, ncol = length(levels(d$chr))))
colnames(boot_mean) <- levels(d$chr)
for (i in levels(d$chr)) {
    d_chr_i <- d %>% filter(chr == i)
    for (n in 1:N) {
        boot_mean[n, i] <- mean(sample_n(d_chr_i, dim(d_chr_i)[1], replace = T)$rho_kb)
    }
}

CI <- t(apply(boot_mean, 2, function(x) quantile(x, c(0.025, 0.975))))
means <- apply(boot_mean, 2, mean)

cbind(d_chr[,c(3)], CI)
```

*The authors state that:*

>“The population-scaled recombination rate is approximately one-third of the mutation rate, consistent with expectations for a species with a high selfing rate.” 

*Using the ratio calculated by the authors (rho_theta) discuss if you support that assertion.*  

Being this statement true, the distribution ratio $\rho/\theta$ should have a mean of $1/3$. Visually:

```{r}
d %>% ggplot() +
    geom_histogram(aes(x = rho_theta, ..density..), bins = 500) +
    geom_vline(xintercept = 1/3, color = "red") +
    geom_text(label = "1/3", x = 0.5, y = 2.5, color = "red")
```


As we have a big sample, a good way of checking if this statement is true would be to make a Confidence Interval for the mean by bootstrap. We will then check whether the value 1/3 falls inside the CI.

```{r}
N = 100
boot_mean <- c()
for (i in levels(d$chr)) {
    boot_mean[i] <- mean(sample(d$rho_theta, length(d$rho_theta), replace = T))
}

quantile(boot_mean, c(0.025, 0.975))
```

According to this, we can't say that the distribution of the ratio $\rho/\theta$ has a mean equal to $1/3$.

Other tests regarding goodness of fit between the distribution of rho and 1/3 of theta, or one sample t-test for the log(rho_theta) (being H0: $\mu = log(1/3)$) lead to the same conclusion.

*Is there a lot of variation between chromosomes for this ratio?*  

We can answer this question performing an ANOVA test.

```{r}
lm <- lm(rho_theta ~ chr, data = d)
anova(lm)
```

As the Analysis of Variance Table returns, there is a lot of variation between chromosomes for this ratio. It may happen that some chromosomes follow this hypothesis and some don't so we will perform the above analysis for each chromosome.

```{r}
d_chr <- d %>% group_by(chr) %>% 
    summarise(mean_rho_theta = mean(rho_theta))

N = 1000
boot_mean <- as.data.frame(matrix(data = 0, nrow = N, ncol = length(levels(d$chr))))
colnames(boot_mean) <- levels(d$chr)
for (i in levels(d$chr)) {
    d_chr_i <- d %>% filter(chr == i)
    for (n in 1:N) {
        boot_mean[n, i] <- mean(sample_n(d_chr_i, dim(d_chr_i)[1], replace = T)$rho_theta)
    }
}

CI <- t(apply(boot_mean, 2, function(x) quantile(x, c(0.025, 0.975))))
cont <- ifelse(CI[, 1] <= 1/3 & CI[, 2] >= 0.3, "*", ".")

data.frame(CI_left = CI[,1], CI_right = CI[, 2], match_hypothesis = cont)
```

We can see that the statement made by the authors is true for chromosomes 3 and 5.

## 4. A test of an evolutionary hypothesis

*One of the conclusion of the paper made by the author is that:*  

>“Nucleotide diversity in 100-kb windows was negatively correlated with gene density, which is expected if diversity is shaped by selection acting against slightly deleterious mutations.“ 

*We want to reproduce analysis and discuss if our analysis support this conclusion?*   

```{r}
cor.test(x = d$gene_dens, y = d$qp.site)

lm <- lm(qp.site ~ gene_dens, data = d)
summary(lm)

d %>% ggplot(aes(x = gene_dens, y = qp.site)) +
    geom_point() +
    geom_smooth(method = "lm")

d %>% ggplot(aes(x = gene_dens, y = qp.site)) +
    facet_wrap(~chr) +
    geom_point() +
    geom_smooth(method = "lm")
```

From the Pearson's correlation test we can see that there is indeed a significant negative correlation between the nucleotide diversity in 100-kb windows (`qp.site` column in data) and the gene density (`gene_dens`). This is also shown in both genome-wide and per chromosome plots.

*The so called “background selection hypothesis” states that genomic regions that are experiencing more deleterious mutations may exhibit overall less polymorphism because selection as it “weeds out” delterious mutations also eliminates other SNP variation in their vicinity. This hypothesis makes the prediction that “everything else being equal”:*  

- *Gene-dense regions (because they are more prone to produce deleterious mutation) should be exhibiting comparatively less polymorphism than regions that are gene poor.*  
- *Background selection will have more influence in regions that have a low recombination rate.*  

*So far testing this hypothesis and its predictions was not easy (except for a few model species) because it requires enormous amounts of polymorphism data. We want to test that hypothesis in M truncatula.*  

**4.1** *Formulate expectations for what genomic covariate will explain variation in nucleotide diversity (qp.site) of a window.*  

First of all, based on the "background selection hypothesis" and the correlation between nucleotide diversity and gene density, we would expect that the gene density will explain variation in nucleotide diversity. Also, the recombination rate is expected to play an important role in the nucleotide diversity. Moreover, we can see there is a high variation between chromosomes for the `qp.site` variable so we could hypothesize that the nucleotide diversity is also dependant on the chromosome. Finally, we could also think that the nucleotide diversity depends on the position of the chromosome.

Therefore, we will perform several correlation test for this variables to determine the best explanation for the variation in nucleotide diversity.

*Using linear models, explore if there is support for the predictions of the background selection hypothesis in the patterns of M truncatula genetic diversity.*  

```{r}

```


**4.2** *Are all chromosome of the M truncatula genome exhibiting similar patterns?*

*(Explore this visually and using appropriate linear models that include a chromosome factor and how chromosome can “interact” with the factors of your anlaysis)*  

Plot the different models in the same plot??

*Can other covariate potentially confounding the background selection hypothesis predictions?*

## 5. Tracking footprints of recent and intense selection in the Medicago truncatula genome.

**5.1** *According to the so-called “selective sweep” scenario, if a region contains a mutation that was recently selected, this recent and intense selection is likely to wipe out most of the polymorphism in that region.*

*Do you find footprints for a recent major selective sweep as evidenced by 2 or more consecutive windows with very low diversity (qp.site) relative to the rest of the genome?*

*Hint: explore graphically where the least polymorphic windows (use the 5% least polymorphic windows within the genome) are located.* 

**5.2** *Is the location of least polymorphic windows randomly distributed on chromosomes?*

*Hint: To make such a test, you can group windows into larger segments comprising 20 windows and count the number of windows containing least polymorphic windows in each segment. “On average” you expect 1 of such 5% least polymorphic windows per segment. Think about what probability distribution you expect can capture the fact that windows with low level of polymorphisms are occurring “at random” along the chromosome in each segment. Then decide on a statistic you can use to test the null hypothesis that the “location of least polymorphic windows is randomly distributed along the chromosomes”. Get a null distribution for such test statistic and state whether you reject the null or not for each chromosome by discussing the p-values you obtain.*

**5.3** *Another typical footprint of a selective sweep is, besides having lower polymorphism, the fact that genomic regions that have experienced a recent selective sweep are expected to have an excess of rare variants (i.e. the only mutations that contribute to polymorphism are recent and therefore still in low frequency). The excess of rare variants in a window can be measured through the summary statistics “Tajima’s D”. Negative Tajima’s D values mean an excess of rare variants (SNPs) while positive Tajima’s D means an excess of frequent SNPs relative to a neutrally evolving region.*

*Are windows exhibiting abnormally low amounts of polymorphism –as measured by qp.site are also having more negative Tajima’s D values?*

*Use some statistical inference method to examine the relationship. Could other confounding variable obscure this relationship?*
